---
title: "sb tag parsing"
date: "`r format(Sys.time(), '%d %B, %Y')`"
params:
  input: "data/processed_sb_values.tsv"
  output: "data/processed_sb_values_summary.tsv"
output: html_document
---
```{r, message=FALSE, warning=FALSE, echo = FALSE, results = FALSE}
library(cowplot)
library(tidyverse)
library(ggsignif)
library(ggokabeito)
library(here)
library(zoo)
library(fs)
library(future)
library(purrr)
library(furrr)
```

Here, we'll do some exploratory data analysis to parse the newly-added 'sb' tags in our bam files, which contain an array of numbers from move tables converted into a samples-per-base metric for each nucleotide using the script 'add_sb_tags.py'.
Have a tsv file from all bams, parsed by cigar string.

```{r}
input_tsv <- params$input
# Read the data in from all mapped reads, all samples
extracted_data <- read_tsv(
  input_tsv,
  col_types = cols(
  read_id = col_character(),
  reference_start = col_double(),
  processed_sb = col_character(),
  file = col_character()
))

extracted_data %>% 
  mutate(processed_sb = strsplit(processed_sb, split = ",") %>% 
           lapply(as.numeric)) %>%
  separate(file, into = c("date", "sample"), extra="drop") -> extracted_data
```
This data is super wide. But to elongate it, we first need a new list-column that iterates from reference_start upwards.
Note: reference still starts at position 0 from python. Will need to update this, but may as well change everything to be relative to the A in CCA as in other plots.
```{r}
extracted_data %>%
  mutate(
    reference_position_list = map2(
      reference_start,
      processed_sb,
      ~ seq(.x, length.out = length(.y))  # Generate a sequence starting at reference_start
    )
  ) %>%
  select(-reference_start) %>%
  unnest(cols = c(processed_sb, reference_position_list)) %>%
  mutate(adjpos = (reference_position_list - 130 + 30 + 1)) ->  prepared_data 
# it's 133 nt to the end of the reference, but account 30 nt of RNA in the 3' adapter,
# plus add 1 to everything because python started ref at 0

rm(extracted_data)
```

```{r}
# what if we took a rolling mean of sb over a 5 nt window for the entire dataset?


prepared_data %>%
  select(-reference_position_list, -date) %>%
  group_by(read_id, sample) %>%
  arrange(adjpos) -> prepared_data

# this is a lot
data_chunks <- prepared_data %>%
  group_split(read_id, sample)

# Set up parallel processing
plan(multisession)

# Define a function to calculate rolling means
calculate_rolling_mean <- function(chunk) {
  chunk %>%
    mutate(sbmean = zoo::rollmean(processed_sb, k = 20, fill = NA, align = "center"))
}

# Apply the function in parallel to each data chunk
result_chunks <- future_map(data_chunks, calculate_rolling_mean)

final_result <- bind_rows(result_chunks)

# let's write this out even though furrr is fassst
write.csv(final_result, params$output, row.names = F)
```

```{r}

AAorder <- c("ala", "asn", "asp", "arg", "cys", "glu", "gln", "gly", "his", "ile", "leu", "lys", "met", "phe", "pro", "ser", "thr", "trp", "tyr", "val", "noAA")

final_result %>%
  ungroup() %>%
  filter(sample != "lys5050mix") %>%
  mutate(sample = factor(sample, levels = AAorder)) %>%
  filter(adjpos == -10) %>%
  ggplot(., aes(y = sbmean, x = sample)) +
  geom_boxplot(outlier.shape = NA) +
  coord_cartesian(ylim = c(0, 300)) +
  theme_minimal_hgrid() +
  theme(aspect.ratio=1) +
  theme(axis.text.x = (element_text(angle = 90, hjust = 1))) +
  labs(title = "Stalling on synthetic tRNA reads",
       x = "",
       y = "Samples per base, \n 20 nt window @ nt -10",
       fill = "") 
```

```{r}
# Calculate means for ordering
order_means <- final_result %>%
  ungroup() %>%
  filter(adjpos == -10) %>% # do this first for memory reasons
  group_by(sample) %>%
  filter(sample != "lys5050mix") %>%
  summarize(mean_sbmean = mean(sbmean, na.rm = TRUE)) %>%
  arrange(mean_sbmean) %>%
  pull(sample)

# plot all reads (this currently lumps multiple runs for a few AAs)
final_result %>%
  ungroup() %>%
  filter(sample != "lys5050mix") %>%
  #filter(adjpos == -10) %>%
  mutate(sample = factor(sample, levels = order_means)) %>%
  ggplot(., aes(y = sbmean, x = sample)) +
  geom_boxplot(outlier.shape = NA) +
  coord_cartesian(ylim = c(0, 300)) +
  theme_minimal_hgrid() +
  theme(aspect.ratio = 1) +
  labs(title = "Stalling on synthetic tRNA reads",
       x = "",
       y = "Samples per base, 20 nt window @ nt -10",
       fill = "") +
  scale_x_discrete(position = "bottom", guide = guide_axis(n.dodge = 1)) +
  theme(axis.text.x = element_text(angle = 90, hjust = 0.5, vjust = 0.4))

```


